### PR TITLE
Fix #79428 - add "$WORKSPACE_FOLDER" for snippet

### LINKED ISSUE #79428: Add WORKSPACE_FOLDER variable to code snippets
Code snippets are great. However, as described in https://code.visualstudio.com/docs/editor/userdefinedsnippets, there's only a build-in variable called `WORKSPACE_NAME` which yields `The name of the opened workspace or folder`, but no variable which yields the full path of the folder opened in VS Code.

Please add another variable called `WORKSPACE_FOLDER` to code snippets, which yields `the path of the folder opened in VS Code`, like the variable of the same name that one can use in `launch.json`and `task.json`, compare https://code.visualstudio.com/docs/editor/variables-reference.

### PR DESCRIPTION
**Pull request by hunghw**
Add a new snippet variable `WORKSPACE_FOLDER`.

### CODE REVIEW

**Review by jrieken — CHANGES_REQUESTED** [review]
please add tests

**Comment by hwhung0111** [comment]
@jrieken I added tests.

File: src/vs/editor/contrib/snippet/test/snippetVariables.test.ts [hunk 1]
```diff
@@ -317,14 +317,21 @@ suite('Snippet Variables Resolver', function () {
...
+			assertVariableResolve(resolver, 'WORKSPACE_FOLDER', '/folderName');
+		}
 
 		// workspace with config
 		const workspaceConfigPath = URI.file('testWorkspace.code-workspace');
 		workspace = new Workspace('', toWorkspaceFolders([{ path: 'folderName' }], workspaceConfigPath), workspaceConfigPath);
 		assertVariableResolve(resolver, 'WORKSPACE_NAME', 'testWorkspace');
+		if (!isWindows) {
+			assertVariableResolve(resolver, 'WORKSPACE_FOLDER', '/');
+		}
```
Comment (jrieken) [inline src/vs/editor/contrib/snippet/test/snippetVariables.test.ts]: why not windows?

File: src/vs/editor/contrib/snippet/snippetVariables.ts [hunk 2]
```diff
@@ -272,13 +274,31 @@ export class WorkspaceBasedVariableResolver implements VariableResolver {
 		}
 
 		if (isSingleFolderWorkspaceIdentifier(workspaceIdentifier)) {
-			return path.basename(workspaceIdentifier.path);
+			if (variable.name === 'WORKSPACE_NAME') {
```
Comment (jrieken) [inline src/vs/editor/contrib/snippet/snippetVariables.ts]: this repeated-if-else is a little ugly, can't you change it be two methods, like `_resolveWorkspaceName` and `_resoveWorkspacePath`?

**Review by jrieken — CHANGES_REQUESTED** [review]

**Review by hunghw — COMMENTED** [review]
@jrieken I separate it into two functions.

File: src/vs/editor/contrib/snippet/test/snippetVariables.test.ts (hunk 1 above)
Comment (jrieken) [reply src/vs/editor/contrib/snippet/test/snippetVariables.test.ts]: 👍

**Pull request merged**